### PR TITLE
Feature/280 import export workflow

### PR DESCRIPTION
**Pull request by jozefbakus**
Implements: #280

### CODE REVIEW

File: ui/src/app/stores/workflows/workflows.effects.spec.ts [hunk 1]
```diff
@@ -23,6 +23,10 @@ import * as WorkflowsActions from './workflows.actions';
 import {
   CreateWorkflow,
   DeleteWorkflow,
+  EXPORT_WORKFLOW_DONE,
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.spec.ts]: Action class, not constant

File: ui/src/app/stores/workflows/workflows.effects.spec.ts [hunk 2]
```diff
@@ -23,6 +23,10 @@ import * as WorkflowsActions from './workflows.actions';
...
+  ExportWorkflow,
+  IMPORT_WORKFLOW_FAILURE,
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.spec.ts]: Action class, not constant

File: ui/src/app/stores/workflows/workflows.effects.spec.ts [hunk 3]
```diff
@@ -814,4 +821,157 @@ describe('WorkflowsEffects', () => {
...
+
+      const action = new ImportWorkflow();
+      mockActions = cold('-a', { a: action });
+
+      const importWorkflowResponse = cold('-#|');
+      spyOn(workflowService, 'importWorkflow').and.returnValue(importWorkflowResponse);
+
+      const expected = cold('--a', {
+        a: {
+          type: IMPORT_WORKFLOW_FAILURE,
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.spec.ts]: I think this should be WorkflowsActions. IMPORT_WORKFLOW_FAILURE not IMPORT_WORKFLOW_FAILURE

File: ui/src/app/stores/workflows/workflows.effects.spec.ts [hunk 4]
```diff
@@ -814,4 +821,157 @@ describe('WorkflowsEffects', () => {
...
+      const blob = new Blob(['hello', ' ', 'world'], { type: 'text/plain' });
+      const response = { blob: blob, fileName: 'fileName' };
+
+      const action = new ExportWorkflow(workflowId);
+      mockActions = cold('-a', { a: action });
+
+      const exportWorkflowResponse = cold('-a|', { a: response });
+      const expected = cold('--a', {
+        a: {
+          type: EXPORT_WORKFLOW_DONE,
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.spec.ts]: I think this should be WorkflowsActions.EXPORT_WORKFLOW_DONE not EXPORT_WORKFLOW_DONE

File: ui/src/app/stores/workflows/workflows.effects.spec.ts [hunk 5]
```diff
@@ -814,4 +821,157 @@ describe('WorkflowsEffects', () => {
...
+
+      const action = new ExportWorkflow(workflowId);
+      mockActions = cold('-a', { a: action });
+
+      const exportWorkflowResponse = cold('-#|');
+      spyOn(workflowService, 'exportWorkflow').and.returnValue(exportWorkflowResponse);
+
+      const expected = cold('--a', {
+        a: {
+          type: EXPORT_WORKFLOW_DONE,
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.spec.ts]: I think this should be WorkflowsActions.EXPORT_WORKFLOW_DONE not EXPORT_WORKFLOW_DONE

File: ui/src/app/stores/workflows/workflows.effects.ts [hunk 6]
```diff
@@ -38,8 +38,9 @@ import { WorkflowHistoriesForComparisonModel, HistoryModel } from '../../models/
 import { WorkflowHistoryService } from '../../services/workflowHistory/workflow-history.service';
 import { JobService } from '../../services/job/job.service';
 import { JobForRunModel } from '../../models/jobForRun.model';
-import { RUN_JOBS } from '../workflows/workflows.actions';
+import { EXPORT_WORKFLOW_DONE } from '../workflows/workflows.actions';
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.ts]: no, not necessary.

File: ui/src/app/stores/workflows/workflows.effects.ts [hunk 7]
```diff
@@ -450,6 +452,81 @@ export class WorkflowsEffects {
...
+          const a = document.createElement('a');
+          a.href = URL.createObjectURL(workflowBlobResponse.blob);
+
+          a.download = workflowBlobResponse.fileName.trim();
+          a.click();
+
+          this.toastrService.success(texts.EXPORT_WORKFLOW_SUCCESS_NOTIFICATION);
+          return [
+            {
+              type: EXPORT_WORKFLOW_DONE,
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.ts]: I think this should be WorkflowsActions.EXPORT_WORKFLOW_DONE not EXPORT_WORKFLOW_DONE

File: ui/src/app/stores/workflows/workflows.effects.ts [hunk 8]
```diff
@@ -450,6 +452,81 @@ export class WorkflowsEffects {
...
+            },
+          ];
+        }),
+        catchError(() => {
+          this.toastrService.error(texts.EXPORT_WORKFLOW_FAILURE_NOTIFICATION);
+          return [
+            {
+              type: EXPORT_WORKFLOW_DONE,
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.ts]: same here

File: ui/src/app/services/workflow/workflow.service.spec.ts [hunk 9]
```diff
@@ -99,6 +99,24 @@ describe('WorkflowService', () => {
     req.flush(new Boolean(true));
   });
 
+  it('exportWorkflow() should return workflow blob', () => {
+    //TODO: Implement test. I need help from you guys.
```
Comment (senelesithole) [inline ui/src/app/services/workflow/workflow.service.spec.ts]: noted

File: ui/src/app/stores/workflows/workflows.effects.ts [hunk 10]
```diff
@@ -450,6 +452,81 @@ export class WorkflowsEffects {
...
+    withLatestFrom(this.store.select(selectWorkflowState)),
+    switchMap(([action, state]: [WorkflowActions.ImportWorkflow, fromWorkflows.State]) => {
+      if (state.workflowAction.workflowFile) {
+        return this.workflowService.importWorkflow(state.workflowAction.workflowFile).pipe(
+          mergeMap((workflow: WorkflowJoinedModel) => {
+            this.toastrService.success(texts.IMPORT_WORKFLOW_SUCCESS_NOTIFICATION);
+            const workflowData = new WorkflowDataModel(workflow, state.workflowAction.workflowFormParts.dynamicParts);
+            return [
+              {
+                type: WorkflowActions.LOAD_WORKFLOW_SUCCESS,
```
Comment (senelesithole) [inline ui/src/app/stores/workflows/workflows.effects.ts:497]: On the first occasion, importing a workflow takes me to create workflow page which is awesome and the workflow is persisted but on the second occasion (importing another workflow) I'm taken to a workflows/import page and the workflow is not persisted.

**Review by senelesithole — CHANGES_REQUESTED** [review]
1. General import fixes.

2. One major functionality issue.

**General Comment**
Importing a non-JSON file error message 'Sorry, something went wrong. Try again.

1. maybe a check first to see if the file uploaded is a JSON object (on modal).
 If not, disable the upload button and alert the user with the specific error
 or let the user click the button but error should be specific.

File: ui/src/app/stores/workflows/workflows.effects.spec.ts (hunk 1 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.spec.ts]: Done

File: ui/src/app/stores/workflows/workflows.effects.spec.ts (hunk 2 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.spec.ts]: Done

File: ui/src/app/stores/workflows/workflows.effects.spec.ts (hunk 3 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.spec.ts]: Done

File: ui/src/app/stores/workflows/workflows.effects.spec.ts (hunk 4 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.spec.ts]: Done

File: ui/src/app/stores/workflows/workflows.effects.spec.ts (hunk 5 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.spec.ts]: Done

File: ui/src/app/stores/workflows/workflows.effects.ts (hunk 6 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.ts]: Done

File: ui/src/app/stores/workflows/workflows.effects.ts (hunk 7 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.ts]: Done

File: ui/src/app/stores/workflows/workflows.effects.ts (hunk 8 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.ts]: Done

File: ui/src/app/stores/workflows/workflows.effects.ts (hunk 10 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.ts:497]: Could you please try it one more time with all commits. I cannot reproduce it.

**Comment by jozefbakus** [comment]
> 1. ror should be specific.

**Comment by jozefbakus** [comment]
> 1. General import fixes.
> 2. One major functionality issue.
> 
> **General Comment**
> Importing a non-JSON file error message 'Sorry, something went wrong. Try again.
> 
> 1. maybe a check first to see if the file uploaded is a JSON object (on modal).
>    If not, disable the upload button and alert the user with the specific error
>    or let the user click the button but error should be specific.

Now only .json files are allowed

**Comment by senelesithole** [comment]
> > 1. General import fixes.
> > 2. One major functionality issue.
> > 
> > **General Comment**
> > Importing a non-JSON file error message 'Sorry, something went wrong. Try again.
> > 
> > 1. maybe a check first to see if the file uploaded is a JSON object (on modal).
> >    If not, disable the upload button and alert the user with the specific error
> >    or let the user click the button but error should be specific.
> 
> Now only .json files are allowed

are you sure? please check again.

**Review by senelesithole — DISMISSED** [review]

**Review by senelesithole — DISMISSED** [review]

**Comment by senelesithole** [comment]
> > 1. General import fixes.
> > 2. One major functionality issue.
> > 
> > **General Comment**
> > Importing a non-JSON file error message 'Sorry, something went wrong. Try again.
> > 
> > 1. maybe a check first to see if the file uploaded is a JSON object (on modal).
> >    If not, disable the upload button and alert the user with the specific error
> >    or let the user click the button but error should be specific.
> 
> Now only .json files are allowed

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/WorkflowController.scala [hunk 11]
```diff
@@ -96,4 +100,24 @@ class WorkflowController @Inject()(workflowService: WorkflowService) {
...
 
+  @GetMapping(path = Array("/workflow/export"))
+  def exportWorkflow(@RequestParam id: Long): CompletableFuture[ResponseEntity[ByteArrayResource]] = {
+    workflowService.getWorkflow(id).map { workflow =>
+      val resource = new ByteArrayResource(
+        ObjectMapperSingleton.getObjectMapper.writeValueAsBytes(workflow)
+      )
+
+      ResponseEntity.ok()
+        .contentType(MediaType.parseMediaType("application/octet-stream"))
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/WorkflowController.scala]: I would use `application/json` here

File: ui/src/app/stores/workflows/workflows.effects.ts [hunk 12]
```diff
@@ -450,6 +450,81 @@ export class WorkflowsEffects {
...
+  workflowExport = this.actions.pipe(
+    ofType(WorkflowActions.EXPORT_WORKFLOW),
+    switchMap((action: WorkflowActions.ExportWorkflow) => {
+      return this.workflowService.exportWorkflow(action.payload).pipe(
+        mergeMap((workflowBlobResponse: { blob: Blob; fileName: string }) => {
+          const a = document.createElement('a');
+          a.href = URL.createObjectURL(workflowBlobResponse.blob);
+
+          a.download = workflowBlobResponse.fileName.trim();
+          a.click();
```
Comment (kevinwallimann) [inline ui/src/app/stores/workflows/workflows.effects.ts:463]: I would add `a.remove();`

File: ui/src/app/services/workflow/workflow.service.spec.ts (hunk 9 above)
Comment (kevinwallimann) [reply ui/src/app/services/workflow/workflow.service.spec.ts]:
```
  it('exportWorkflow() should return workflow blob', () => {
    const content = '{"workflowId":"1"}';
    const blob = new Blob([content], { type: 'application/json' });
    const filename = 'filename.json';
    const id = 1;
    underTest.exportWorkflow(id).subscribe(
      (data) => {
        expect(data.fileName).toEqual(filename);
        expect(data.blob).toEqual(blob);
      },
      (error) => fail(error),
    );
    const req = httpTestingController.expectOne(api.EXPORT_WORKFLOW + `?id=${id}`);
    expect(req.request.method).toEqual('GET');
    req.flush(blob, {
      headers: { 'Content-Disposition': `attachment; filename=${filename}` },
    });
  });


```

Comment (jozefbakus) [reply ui/src/app/services/workflow/workflow.service.spec.ts]: Done, thanks

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/WorkflowController.scala (hunk 11 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/WorkflowController.scala]: Done

File: ui/src/app/stores/workflows/workflows.effects.ts (hunk 12 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.ts:463]: Done

**Review by kevinwallimann — APPROVED** [review]

**Pull request merged**